modules/woocommerce-analytics/classes/wp-woocommerce-analytics-universal.php

```diff
@@ -0,0 +1,198 @@
+<?php
+
+/**
+* Jetpack_WooCommerce_Analytics_Universal
+*
+* @author greenafrican
+*/
+
+/**
+* Bail if accessed directly
+*/
+if ( ! defined( 'ABSPATH' ) ) {
+	exit;
+}
+
+class Jetpack_WooCommerce_Analytics_Universal {
+	public function __construct() {
+		add_action( 'woocommerce_add_to_cart', array( $this, 'add_to_cart' ) );
+		add_action( 'wp_head', array( $this, 'wp_head' ), 999999 );
+		add_action( 'wp_footer', array( $this, 'loop_add_to_cart' ) );
+		add_action( 'woocommerce_after_cart', array( $this, 'remove_from_cart' ) );
+		add_action( 'woocommerce_after_mini_cart', array( $this, 'remove_from_cart' ) );
+		add_filter( 'woocommerce_cart_item_remove_link', array( $this, 'remove_from_cart_attributes' ), 10, 2 );
+		add_action( 'woocommerce_after_single_product', array( $this, 'product_detail' ) );
+		// add_action( 'woocommerce_after_checkout_form', array( $this, 'checkout_process' ) );
+	}
+
+	public function wp_head() {
+
+		// If we're in the admin_area, return without inserting code.
+		if ( is_admin() ) {
+			return;
+		}
+
+		if ( ! class_exists( 'WooCommerce' ) ) {
+			return;
+		}
+
+		$async_code = "<script async src='https://stats.wp.com/s.js'></script>";
+
+		echo "$async_code\r\n";
+	}
+
+	public function add_to_cart() {
+
+		if ( ! is_single() ) {
+			return;
+		}
+		$blogid = Jetpack::get_option( 'id' );
+		global $product;
+
+		$product_sku_or_id = Jetpack_WooCommerce_Analytics_Utils::get_product_sku_or_id( $product );
+		$selector = ".single_add_to_cart_button";
+
+		wc_enqueue_js(
+			"window._wca = window._wca || [];
+			$( '" . esc_js( $selector ) . "' ).click( function() {
+				_wca.track( {
+					'_en': 'woocommerce_analytics_add_to_cart',
+					'blog_id': " . esc_js( $blogid ) . ",
+					'pi': '" . esc_js( $product_sku_or_id ) . "',
+					'pn' : '" . esc_js( $product->get_title() ) . "',
+					'pq': $( 'input.qty' ).val() ? $( 'input.qty' ).val() : '1',
+				} );
+			)};"
+		);
+	}
+
+	public function loop_add_to_cart() {
+
+		if ( ! class_exists( 'WooCommerce' ) ) {
+			return;
+		}
+
+		$minimum_woocommerce_active = class_exists( 'WooCommerce' ) && version_compare( WC_VERSION, '3.0', '>=' );
+		if ( ! $minimum_woocommerce_active ) {
+			return;
+		}
+		$blogid = Jetpack::get_option( 'id' );
+		$selector = ".add_to_cart_button:not(.product_type_variable, .product_type_grouped)";
+
+		wc_enqueue_js(
+			"$( '" . esc_js( $selector ) . "' ).click( function() {
```

Comment:
By listening to the `added_to_cart` Woo event, we can fire our added to cart event when the server returns without errors on the callback.
```js
jQuery('body').on('added_to_cart',function(){
    // call track event here
}
```
Is there an advantage to sending the event as soon as the user clicks the button?

Reply:
The same situation for "remove from cart"

Reply:
Except that event isn't fired for the non-async add-to-cart, is it?

Reply:
That right. The `add_to_cart` function above still has to handle non-async